src/tools/tidy/src/fluent_alphabetical.rs

```diff
@@ -0,0 +1,67 @@
+//! Checks that all Flunt files have messages in alphabetical order
+
+use crate::walk::walk;
+use std::{fs::OpenOptions, io::Write, path::Path};
+
+use regex::Regex;
+
+lazy_static::lazy_static! {
+    static ref MESSAGE: Regex = Regex::new(r#"(?m)^([a-zA-Z0-9_]+)\s*=\s*"#).unwrap();
+}
+
+fn filter_fluent(path: &Path) -> bool {
+    if let Some(ext) = path.extension() { ext.to_str() != Some("ftl") } else { true }
+}
+
+fn check_alphabetic(filename: &str, fluent: &str, bad: &mut bool) {
+    let mut matches = MESSAGE.captures_iter(fluent).peekable();
+    while let Some(m) = matches.next() {
+        if let Some(next) = matches.peek() {
+            let name = m.get(1).unwrap();
+            let next = next.get(1).unwrap();
+            if name.as_str() > next.as_str() {
+                tidy_error!(
+                    bad,
+                    "{filename}: message `{}` appears before `{}`, but is alphabetically later than it
+run tidy with `--bless` to sort the file correctly",
+                    name.as_str(),
+                    next.as_str()
+                );
+            }
+        } else {
+            break;
+        }
+    }
+}
+
+fn sort_messages(fluent: &str) -> String {
+    let mut chunks = vec![];
+    let mut cur = String::new();
+    for line in fluent.lines() {
+        if MESSAGE.is_match(line) {
+            chunks.push(std::mem::take(&mut cur));
+        }
+        cur += line;
+        cur.push('\n');
+    }
+    chunks.push(cur);
+    chunks.sort();
+    let mut out = chunks.join("");
+    out = out.trim().to_string();
+    out.push('\n');
+    out
+}
+
+pub fn check(path: &Path, bless: bool, bad: &mut bool) {
+    walk(path, |path, is_dir| !is_dir && filter_fluent(path), &mut |ent, contents| {
```

Comment:
This looks wrong, it won't recurse into any subdirectory. Also it should use the default `filter_dirs` function to avoid changing clippy.

Reply:
It worked in my testing, erroring on misordered files and blessing them. As far as I understand, this will skip only files that aren't `.ftl` ones
Clippy isn't affected here since we only pass `compiler_path` in main, but I'll add that filter to make sure it continues to work correctly

Reply:
oh I forgot the api, this returns true if it *should* recurse. yeah this code looks fine, sorry for the noise.